### PR TITLE
#0: Remove alignment requirements for Row Major tensors

### PR DESCRIPTION
**Pull request by sminakov-tt**
### Problem description
Currently its required that RowMajor tensor size must be aligned to 4 bytes, because the data is transferred to device as a vector of `uint32_t`. This creates limitations in TTNN and complicates handling of Row Major tensors.
This PR is a prerequisite for https://github.com/tenstorrent/tt-metal/pull/15028
Thanks to @abhullar-tt for helping with lifting those restrictions.

### What's changed
Modify Metal API allowing to send an arbitrary vector to device
Modify PageConfig removing alignment restrictions
Remove conversions to/from `vector<uint32_t>` for sending tensor to/from device
Remove asserts on buffer alignment for row major
Enable previously disabled tensor tests

### Checklist
- [x] [Post commit CI passes](https://github.com/tenstorrent/tt-metal/actions/runs/11944863653)
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/11929076543)
- [x] [Model regression CI testing passes](https://github.com/tenstorrent/tt-metal/actions/runs/11929113643)
- [x] [Device performance regression CI testing passes](https://github.com/tenstorrent/tt-metal/actions/runs/11929096855)
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: tests/ttnn/unit_tests/gtests/tensor/test_create_tensor.cpp [hunk 1]
```diff
@@ -121,9 +121,9 @@ INSTANTIATE_TEST_SUITE_P(
     EmptyTensorTest,
     ::testing::Combine(
         ::testing::Values(
-            //ttnn::Shape({}),
-            //ttnn::Shape({0}),
-            //ttnn::Shape({1}),
+            ttnn::Shape({}),
+            ttnn::Shape({0}),
+            ttnn::Shape({1}),
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/gtests/tensor/test_create_tensor.cpp:126]: great

File: tests/ttnn/unit_tests/gtests/tensor/test_sharding_with_alignment.cpp [hunk 2]
```diff
@@ -822,10 +822,10 @@ INSTANTIATE_TEST_SUITE_P(
         },
         /////////////////////////////////////////////////////////////////////////////////////////////////////
         // EXAMPLE 2: ROW_MAJOR tensor with different representation for width sharded / interleaved
-        // - In this example, (shard) width alignment is 4 because UINT8 = 1 bytes and we pack with uint32_t
+        // - In this example, (shard) width alignment is 1 because it's row major
         /////////////////////////////////////////////////////////////////////////////////////////////////////
         // Example 2a: Logical shard shape + alignment after
-        // - Along width: 5 / 1 is 5 shards; 5 * 4 = 20
+        // - Along width: 5 / 1 is 5 shards; 5 * 1 = 5
```
Comment (ayerofieiev-tt) [inline tests/ttnn/unit_tests/gtests/tensor/test_sharding_with_alignment.cpp:828]: 🤣

Comment (ayerofieiev-tt) [reply tests/ttnn/unit_tests/gtests/tensor/test_sharding_with_alignment.cpp:828]: @TT-BrianLiu ptal

File: tt_metal/impl/dispatch/memcpy.hpp [hunk 3]
```diff
@@ -23,7 +23,6 @@ using vector_memcpy_aligned = std::vector<T, tt::stl::aligned_allocator<T, MEMCP
 template <bool debug_sync = false>
 static inline void memcpy_to_device(void *__restrict dst, const void *__restrict src, size_t n) {
```
Comment (ayerofieiev-tt) [inline tt_metal/impl/dispatch/memcpy.hpp:24]: anybody knows why this code exists? why is it not just `std::memcpy`?

File: ttnn/cpp/ttnn/tensor/tensor_impl.hpp [hunk 4]
```diff
@@ -51,112 +51,6 @@ std::vector<OutputDataType> cast_vec(const BufferType<InputDataType>& data_to_co
     return converted_data;
 }
 
-// TODO(arakhmati): Should pack_vec_into_uint32_vec be a generator?
-template <typename DataType, template <typename> typename BufferType>
-std::vector<uint32_t> pack_vec_into_uint32_vec(const BufferType<DataType>& data_to_pack) {
```
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/tensor/tensor_impl.hpp:56]: ha, nice.

this was basically replaced with reinterpret_cast?

**Review by ayerofieiev-tt — APPROVED** [review]
1. Major problem solved
2. Complexity reduced
3. LoC removed

I hope this works perfectly well for all architectures.

File: tt_metal/impl/dispatch/memcpy.hpp (hunk 3 above)
Comment (tt-aho) [reply tt_metal/impl/dispatch/memcpy.hpp:24]: For performance 🚀

File: tt_metal/llrt/tt_cluster.hpp [hunk 5]
```diff
@@ -91,6 +91,8 @@ class Cluster {
         void *mem_ptr, uint32_t sz_in_bytes, tt_cxy_pair core, uint64_t addr, bool small_access = false) const;
     void read_core(
         std::vector<uint32_t> &data, uint32_t sz_in_bytes, tt_cxy_pair core, uint64_t addr, bool small_access = false) const;
+    void read_core(
```
Comment (abhullar-tt) [inline tt_metal/llrt/tt_cluster.hpp:94]: same thing here, don't think this is needed since there are read_core variants that accept void*

**Review by abhullar-tt — COMMENTED** [review]
Thanks for lifting a constraint that was only applicable to slow dispatch API limitation :). Overall lgtm but some minor clean up requested

File: tt_metal/impl/dispatch/memcpy.hpp (hunk 3 above)
Comment (ayerofieiev-tt) [reply tt_metal/impl/dispatch/memcpy.hpp:24]: @tt-aho who is the best person to share more about it? Usually such code might be faster in a specific context, sometimes enabling specific compiler optimizations will produce a better result across platforms.

File: tt_metal/impl/dispatch/memcpy.hpp [hunk 6]
```diff
@@ -68,6 +67,12 @@ static inline void memcpy_to_device(void *__restrict dst, const void *__restrict
                 src8 += sizeof(int32_t);
                 dst8 += sizeof(int32_t);
             }
+            n -= n / sizeof(int32_t) * sizeof(int32_t);
+            for (size_t i = 0; i < n; ++i) {
```
Comment (Denys88) [inline tt_metal/impl/dispatch/memcpy.hpp:71]: is it memcpy?

File: tt_metal/impl/dispatch/memcpy.hpp (hunk 3 above)
Comment (dmakoviichuk-tt) [reply tt_metal/impl/dispatch/memcpy.hpp:24]: do you have any tests which shows that it is faster then memcpy?

Comment (tt-aho) [reply tt_metal/impl/dispatch/memcpy.hpp:24]: We have some tests here: `tests/tt_metal/tt_metal/perf_microbenchmark/3_pcie_transfer/test_pull_from_pcie.cpp`

@pgkeller @tt-asaigal for reference regarding usage of avx/non-temporal instructions

File: tt_metal/impl/dispatch/memcpy.hpp (hunk 6 above)
Comment (tt-aho) [reply tt_metal/impl/dispatch/memcpy.hpp:71]: @pgkeller I recall you mentioning we'd always need to use non-temporal stores here and can't switch between them

File: tests/ttnn/unit_tests/gtests/tensor/test_sharding_with_alignment.cpp (hunk 2 above)
Comment (TT-BrianLiu) [reply tests/ttnn/unit_tests/gtests/tensor/test_sharding_with_alignment.cpp:828]: looks good now

File: tests/ttnn/unit_tests/gtests/tensor/test_sharding_with_alignment.cpp [hunk 7]
```diff
@@ -822,10 +822,10 @@ INSTANTIATE_TEST_SUITE_P(
         },
         /////////////////////////////////////////////////////////////////////////////////////////////////////
         // EXAMPLE 2: ROW_MAJOR tensor with different representation for width sharded / interleaved
-        // - In this example, (shard) width alignment is 4 because UINT8 = 1 bytes and we pack with uint32_t
+        // - In this example, (shard) width alignment is 1 because it's row major
         /////////////////////////////////////////////////////////////////////////////////////////////////////
         // Example 2a: Logical shard shape + alignment after
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/gtests/tensor/test_sharding_with_alignment.cpp:827]: Can you change it so it has same comment as example 2b: 
```
        // Example 2a: Logical shard shape that is already aligned
        // NOTE: ShardMode::PHYSICAL is equivalent in this case
```

File: tests/ttnn/unit_tests/gtests/tensor/test_sharding_with_alignment.cpp [hunk 8]
```diff
@@ -844,15 +844,15 @@ INSTANTIATE_TEST_SUITE_P(
...
         // Example 2b: Logical shard shape that is already aligned
         // NOTE: ShardMode::PHYSICAL is equivalent in this case
-        // - Along width: 5 / 4 is 2 shards; 2 * 4 = 8
+        // - Along width: 8 / 4 is 2 shards; 2 * 4 = 8
         CreateShardedTensorWithAlignmentParams{
             CreateShardedTensorWithAlignmentInputs{
-                .shape = SimpleShape{1, 2, 10, 5},
+                .shape = SimpleShape{1, 2, 10, 8},
                 .data_type = DataType::UINT8,
                 .page_config = PageConfig(Layout::ROW_MAJOR),
```
Comment (TT-BrianLiu) [inline tests/ttnn/unit_tests/gtests/tensor/test_sharding_with_alignment.cpp:857]: Why change 5 to 8? We support partial shards. Can you keep old specs for this test?

**Review by TT-BrianLiu — COMMENTED** [review]
We must also be adding padding somewhere when creating row major tensors with previously unsupported dims right? Is the next step to remove this padding?

File: ttnn/cpp/ttnn/tensor/tensor_impl.hpp (hunk 4 above)
Comment (sminakov-tt) [reply ttnn/cpp/ttnn/tensor/tensor_impl.hpp:56]: Yes, there were a lot of op tests which went through this flow

File: tt_metal/detail/tt_metal.hpp [hunk 9]
```diff
@@ -56,12 +76,22 @@ inline namespace v0 {
...
         * | shard_order | For a sharded buffer we can read in shard order | bool                    |                                                  | No       |
         */
-        void ReadFromBuffer(Buffer &buffer, std::vector<uint32_t> &host_buffer, bool shard_order = false);
-        void ReadFromBuffer(std::shared_ptr<Buffer> buffer, std::vector<uint32_t> &host_buffer, bool shard_order = false);
-
+        template<typename DType>
+        void ReadFromBuffer(Buffer &buffer, std::vector<DType> &host_buffer, bool shard_order = false) {
+            auto buffer_size = buffer.size();
+            TT_FATAL(buffer_size % sizeof(DType) == 0, "Buffer size is not divisible by dtype size");
+            host_buffer.resize(buffer.size() / sizeof(DType));
```
[the remark this answers is not in the document]
Comment (sminakov-tt) [reply tt_metal/detail/tt_metal.hpp:86]: It is needed. The expectation that if read function takes a vector it'll resize it appropriately

File: tt_metal/impl/dispatch/memcpy.hpp (hunk 3 above)
Comment (sminakov-tt) [reply tt_metal/impl/dispatch/memcpy.hpp:24]: @tt-aho @pgkeller @tt-asaigal @ayerofieiev-tt @dmakoviichuk-tt  I ran `test_pull_from_pcie` and I observe identical perf values of this code vs `std::memcpy`(I'm using Ryzen 9 7950X3D and a single N150 board).

Let's create a separate ticket to take a look at this, but leave it as-is in this PR :)

Comment (tt-aho) [reply tt_metal/impl/dispatch/memcpy.hpp:24]: On the T3Ks we have I think we were seeing ~6GB/s with `std::memcpy`, and ~10GB/s with  `memcpy_to_device`, but we haven't really looked into the perf for this recently. It was a noticeable gain when the custom implementation was originally added.

**Review by abhullar-tt — APPROVED** [review]

**Comment by sminakov-tt** [comment]
@TT-BrianLiu We will have a general effort of removing all extra paddings throughout the codebase for both Tile and RowMajor once we get this PR and TensorLayout in Tensor merged

**Review by tt-aho — COMMENTED** [review]
Changes look okay to me but please wait for confirmation/resolution of this thread before merge https://github.com/tenstorrent/tt-metal/pull/15245#discussion_r1850726169

File: tt_metal/impl/dispatch/memcpy.hpp (hunk 3 above)
Comment (sminakov-tt) [reply tt_metal/impl/dispatch/memcpy.hpp:24]: Filed an issue for investigation https://github.com/tenstorrent/tt-metal/issues/15275
Closing this conversation as not related to this PR.

**Comment by sminakov-tt** [comment]
@tt-aho The conversation about `memcpy` is irrelevant to this PR. The PR just adds handling on `n % 4 != 0` into the same function in the same paradigm. I created an issue for re-evaluation of `memcpy_to_device`: https://github.com/tenstorrent/tt-metal/issues/15275. 

Let's move forward with this PR

**Comment by tt-aho** [comment]
@sminakov-tt the conversation I linked to is relevant, as the change to memcpy_to_device may cause issues due to mixing temporal and non-temporal stores to hugepage.

**Comment by sminakov-tt** [comment]
@tt-aho Thanks for clarifying! @pgkeller Can you elaborate more on temporal /non-temporal stores?

File: tt_metal/impl/dispatch/memcpy.hpp (hunk 6 above)
Comment (pgkeller) [reply tt_metal/impl/dispatch/memcpy.hpp:71]: right.  if you do a store, the data will be pulled into cache and you then run the risk of a cacheline flush later corrupting data.  I think w/ the pcie device being cache coherent that that may solve the issue (device pulls from memory, flushing the cache) but I'm not sure if the flush is an eviction or just a write or if that behavior is defined or not.

ideally, these writes would take aligned buffers and we'd control the allocation to ensure they are always aligned, then the copies would always use si128 writes (even potentially writing off the end by a few bytes)

**Comment by pgkeller** [comment]
> @tt-aho Thanks for clarifying! @pgkeller Can you elaborate more on temporal /non-temporal stores?

discussed w/ @sminakov-tt , data can be stored to stack and copied to destination

**Comment by cfjchu** [comment]
any reason to prefer uint8_t over std::byte?

**Review by cfjchu — APPROVED** [review]
Looks great!

**Comment by sminakov-tt** [comment]
@cfjchu There's no specific reason, `uint8_t` seems to be used more often in the codebase

**Review by tt-aho — APPROVED** [review]

**Pull request merged**